### PR TITLE
#948(NH-3435) - Ability to select entities in a criteria projections (rebased PR #302)

### LINKED ISSUE #948: NH-3435 - Ability to select the root entity in a criteria projection
**Jonathan Hynes** created an [issue](https://nhibernate.jira.com/browse/NH-3435) — <time datetime="2013-04-01T13:44:03.748Z" title="2013-04-01T13:44:03.748Z">1st April 2013, 13:44:03</time>:
> The Criteria API need a way to select an Entity as a projection. Suggest the following methods:
> ```c#
> Projection.RootEntity()
> Projection.Entity(string associationPath)
> ```
> This issue may be a duplicate of one of the following, but these issues have been closed and the problem is not resolved:
> [NH-1372](https://nhibernate.jira.com/browse/NH-1372)
> [NH-928](https://nhibernate.jira.com/browse/NH-928)
> 
> Background to the issue can be found here:
> http://stackoverflow.com/questions/15744057/nhibernate-how-to-select-the-root-entity-in-a-projection
> 
---
**Ricardo Peres** added a comment — <time datetime="2014-03-06T20:46:57.344Z" title="2014-03-06T20:46:57.344Z">6th March 2014, 20:46:57</time>:
> I implemented a solution: http://weblogs.asp.net/ricardoperes/archive/2014/03/06/custom-nhibernate-criteria-projections.aspx.
> If anyone can confirm that it works, I may submit a pull request.
> 
---
**Jonathan Hynes** added a comment — <time datetime="2014-03-14T22:42:07.238Z" title="2014-03-14T22:42:07.238Z">14th March 2014, 22:42:07</time>:
> I tested using the `MsSql2008Dialect`. This does not appear to work because aliases are the property name and are not escaped/quoted. This can cause issues if keywords are used as column names. This could also cause problems if multiple projections have the same property name.
> 
---
**Ricardo Peres** added a comment — <time datetime="2014-03-14T22:56:53.768Z" title="2014-03-14T22:56:53.768Z">14th March 2014, 22:56:53</time>:
> What does "does not appear to work" mean? Does it return, in normal cases, what it should, or does it not?
> Not using aliases was a deliberate option, so that we can use `AliasToBean` in order to generate entities from the projection. Quoting is easy to do.
> Anyway, this was a proof of concept, I even made chances to the sample after I published that post. If I ever submit a pull request, I will address some of this issues.
> 
---
**Jonathan Hynes** added a comment — <time datetime="2014-03-14T23:11:03.263Z" title="2014-03-14T23:11:03.263Z">14th March 2014, 23:11:03</time>:
> My test is as follows:
> ```c#
> //Code to persist 20 Person objects here
> 
> var list = Session.CreateCriteria<Person>()
>               .SetFirstResult(5)
>               .SetMaxResults(10)
>               .SetProjection(new RootEntityProjection(), Projections.SqlFunction("rowcount", NHibernateUtil.Int64))
>               .List<object[]>();
> 
> Assert.Equal(10, list.Count);
>             Assert.Equal(20, (long)list<0>[1]);
>             Assert.IsType<Person>(list<0>[0]);
> ```
> The `Person` class has a `User` property. Since `User` is a SQL server keyword, I am getting the following error:
> ```
> System.Data.SqlClient.SqlException : Incorrect syntax near the keyword 'User'.
> ```
> Looking at the generated SQL, given the lack of uniquifying aliases, I also make the conjecture that the generated SQL could have alias conflicts. For instance, suppose `Person` had the following property:
> ```c#
> public virtual Person Father{get;set;}
> ```
> And I modified the query to say:
> ```c#
> .CreateAlias("Father", "f")
> .SetProjection(new RootEntityProjection(), new EntityProjection(Person, "f"))
> ```
> I conjecture that this would create duplicate aliases in the generated SQL for every property in the `Person` class.
> 
---
**Ricardo Peres** added a comment — <time datetime="2014-03-14T23:18:08.259Z" title="2014-03-14T23:18:08.259Z">14th March 2014, 23:18:08</time>:
> The problem with having aliases is that we cannot turn the results into entities. Anyway, I will post a new version that addresses your concerns.
> 
---
**Ricardo Peres** added a comment — <time datetime="2014-03-15T11:22:15.114Z" title="2014-03-15T11:22:15.114Z">15th March 2014, 11:22:15</time>:
> Jonathan, thanks for your feedback! Can I please ask you to try this new version?
> 
> 
---
**Jonathan Hynes** added a comment — <time datetime="2014-03-15T12:07:05.013Z" title="2014-03-15T12:07:05.013Z">15th March 2014, 12:07:05</time>:
> Now I am getting this error:
> ```
> System.ArgumentOutOfRangeException : Index was out of range. Must be non-negative and less than the size of the collection.
> Parameter name: index
> Result StackTrace:	
> at System.ThrowHelper.ThrowArgumentOutOfRangeException()
>    at System.SZArrayHelper.get_Item<T>(Int32 index)
>    at System.Linq.Enumerable.ElementAt<TSource>(IEnumerable`1 source, Int32 index)
>    at NHibernate.Criterion.BaseProjection.NHibernate.Criterion.IProjection.ToSqlString(ICriteria criteria, Int32 position, ICriteriaQuery criteriaQuery, IDictionary`2 enabledFilters) in d:\dev\npf\Drg.M3\Dao\Base\BaseProjection.cs:line 131
>    at NHibernate.Criterion.ProjectionList.ToSqlString(ICriteria criteria, Int32 loc, ICriteriaQuery criteriaQuery, IDictionary`2 enabledFilters)
>    at NHibernate.Loader.Criteria.CriteriaQueryTranslator.GetSelect(IDictionary`2 enabledFilters)
>    at NHibernate.Loader.Criteria.CriteriaJoinWalker..ctor(IOuterJoinLoadable persister, CriteriaQueryTranslator translator, ISessionFactoryImplementor factory, ICriteria criteria, String rootEntityName, IDictionary`2 enabledFilters)
>    at NHibernate.Loader.Criteria.CriteriaLoader..ctor(IOuterJoinLoadable persister, ISessionFactoryImplementor factory, CriteriaImpl rootCriteria, String rootEntityName, IDictionary`2 enabledFilters)
>    at NHibernate.Impl.SessionImpl.List(CriteriaImpl criteria, IList results)
>    at NHibernate.Impl.CriteriaImpl.List(IList results)
>    at NHibernate.Impl.CriteriaImpl.List<T>()
>    at Drg.M3.Tests.Unit.Core.NHibernateFacts.RootEntityProjectionWorks() in d:\dev\npf\Drg.M3.Tests.Unit\Core\NHibernateFacts.cs:line 26
> ```
> FYI, my persister is a `JoinedSubclassEntityPersister`. It's getting this error on the first property and first column name (i==0, c==0), and the property in question is a one-to-one.
> 
---
**Ricardo Peres** added a comment — <time datetime="2014-08-13T14:31:12.201Z" title="2014-08-13T14:31:12.201Z">13th August 2014, 14:31:12</time>:
> Pull request: https://github.com/nhibernate/nhibernate-core/pull/299
> 
---
**Ricardo Peres** added a comment — <time datetime="2014-08-14T10:22:41.085Z" title="2014-08-14T10:22:41.085Z">14th August 2014, 10:22:41</time>:
> New pull request: https://github.com/nhibernate/nhibernate-core/pull/302
> 
---
**Ricardo Peres** added a comment — <time datetime="2014-10-10T13:26:18.398Z" title="2014-10-10T13:26:18.398Z">10th October 2014, 13:26:18</time>:
> Fix version? Maybe 4.1?
> 
---
**Alexander Zaytsev** added a comment — <time datetime="2016-08-12T04:40:11.455Z" title="2016-08-12T04:40:11.455Z">12th August 2016, 4:40:11</time>:
> Move unresolved improvements and new features with minor and trivial priorities to 4.2.0
> 
---
**Alexander Zaytsev** added a comment — <time datetime="2017-09-14T01:47:17.339Z" title="2017-09-14T01:47:17.339Z">14th September 2017, 1:47:17</time>:
> Move issues from 4.2 to 5.0
> 
---
**Alexander Zaytsev** added a comment — <time datetime="2017-09-14T01:50:03.693Z" title="2017-09-14T01:50:03.693Z">14th September 2017, 1:50:03</time>:
> Move new features to 5.1
>

### PR DESCRIPTION
**Pull request by bahusoid**
Fixes #948.
This PR allows to select  entities (lazy/eagerly) via Criteria API (via `Projections.Entity` and `Projections.RootEntity` helper methods).

I rebased #302 pull request +
1) Added more tests (lazy and eager version);
2) Refactored BaseEntityProjection to use `ILoadable` interface instead of `AbstractEntityPersister`;
3) Extended Projections helper class to expose lazy setting and added helper for QueryOver.

Thanks Ricardo Peres for his work!

### CODE REVIEW

**Comment by bahusoid** [comment]
Sorry. Somehow I messed up with tests in final commit.... Working on the fix..

**Comment by bahusoid** [comment]
And just realized that eager  version is not really eager - criteria leads to 2 selects ID only and full entity... ( And fixing it doesn't look easy...

**Comment by fredericDelaporte** [comment]
#302/#948 were about choosing the root entity to load. Does this point works as expected? (Loads fully the desired entity, not just uninitialized proxies.) If yes, better re-target this PR at just that. And maybe open new issues/pr for the eager fetching issues. (Check before if none are already existing for this other subject.)

Test failures for Postgres, Oracle and Firebird are all alike: for some reason, it seems it generates bad column names, alias names or badly combine both.

**Comment by bahusoid** [comment]
@fredericDelaporte But is it really useful? 
The root of the issue described in https://nhibernate.jira.com/browse/NH-3435 based on stackoverflow question http://stackoverflow.com/questions/15744057/nhibernate-how-to-select-the-root-entity-in-a-projection.

Both JIRA issue and stackoverflow question by the same author.

And the problem he tries to solve -  implement Criteria API variant of the following HQL query:
```
session.CreateQuery("select b, rowcount() from Blog b")
              .SetFirstResult(5)
              .SetMaxResults(10)
              .List();
```
And HQL does load entities eagerly.

Also see comment in sackoverflow question:
>Yes, I know I could just use CriteriaTransform.TransformToRowCount() but that would require executing the query twice - once for the results and once for the row count. Using Futures may help a little by reducing it to one round-trip, but it's still executing the query twice on the SQL Server. For intensive queries this is unacceptable. **I want to avoid the overhead, and return the row count and the results in the same query.**

So I would say issue was created for supporting projections for fully initialized entities.

And if I understand correctly @rjperes  first [comment](https://nhibernate.jira.com/browse/NH-3435?focusedCommentId=32052&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-32052). His first implementation was exactly about eager loading. And for some reasons final proposed implementation transformed into lazy loading.

 IMHO standard implementation if exists should support eager loading . Otherwise it's not very useful.

**Comment by fredericDelaporte** [comment]
Ok, so the issue (the migrated one on GitHib, #948) should be updated for better matching this. May you add a comment redefining the issue? I would then update its title and first comment.

**Comment by bahusoid** [comment]
@fredericDelaporte It seems we have some kind of misunderstanding (maybe I'm using wrong terminology).

> #302/#948 were about choosing the root entity to load. Does this point works as expected? (Loads fully the desired entity, not just uninitialized proxies.)

Current implementation doesn't fully load entities (no matter root or not) - it loads uninitialized  proxies instead (only ID column retrieved from database).

**Comment by fredericDelaporte** [comment]
Well, there is some mis-understanding. I would not use "eager loading" term for the case where we select an entity along with some additional data (the rowcount here). From what you write, I now understand we happen to have a bug in this PR because instead of actually selecting the entities it just returns uninitialized proxies. Then for me such a point has to be fixed in this PR, and we should not need to specify lazy false or anything like that for having the entity fully loaded by the this query.

When I see "eager loading", I am thinking about related entities, say by example having also a `rootEntity.SomeRelatedEntity` loaded along. If such a case was also needing some work, then it should likely have its own PR.

Now I have to say I am not using Criteria much, so my view may not be the right one.

**Comment by bahusoid** [comment]
As I suspected retrieving initialized entities wasn't an easy fix. I made a draft with such support and there is not much left from initial Ricardo's implementation.

@fredericDelaporte , @hazzik  Could you please review specifically my latest [commit](https://github.com/nhibernate/nhibernate-core/pull/1458/commits/afb281c71d1f65433df0f16289e99a2933f6e76e). I temporarily isolated changes in test solution (see [EntityProjection.cs](https://github.com/nhibernate/nhibernate-core/pull/1458/commits/afb281c71d1f65433df0f16289e99a2933f6e76e#diff-689366c9d925cc6057310a1e0ae0227e) and [EntityProjectionType.cs](https://github.com/nhibernate/nhibernate-core/pull/1458/commits/afb281c71d1f65433df0f16289e99a2933f6e76e#diff-da9d86bb3235599a75b3aea957dc98cb). I also temporarily modified test csproj to include only related tests (new feature doesn't affect any existing functionality - and I hope it will make tests executed on server much quicker)

File: src/NHibernate.Test/NHSpecificTest/GH948/EntityProjection.cs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+			{
+				RootEntity = criteria.GetRootEntityTypeIfAvailable();
+			}
+
+			if (_entityAlias == null)
+			{
+				_entityAlias = criteria.Alias;
+			}
+			
+			Persister = (IQueryable) criteriaQuery.Factory.GetEntityPersister(RootEntity.FullName);
```
Comment (bahusoid) [inline src/NHibernate.Test/NHSpecificTest/GH948/EntityProjection.cs]: Is this valid cast? Any more proper way to obtain `IQueryable`

File: src/NHibernate.Test/NHSpecificTest/GH948/EntityProjectionType.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+				identifier,
+				_projection.PropertyAliases,
+				LockMode.None,
+				_projection.FetchLazyProperties,
+				_projection.IsReadOnly);
+
+			return entity;
+		}
+
+		private static object CreateInitializedEntity(DbDataReader rs, ISessionImplementor session, IQueryable persister, object identifier, string[][] propertyAliases, LockMode lockMode, bool fetchLazyProperties, bool readOnly)
```
Comment (bahusoid) [inline src/NHibernate.Test/NHSpecificTest/GH948/EntityProjectionType.cs]: The main concern is this method. If/when it's properly implemented I think it should be moved to some utility class. As finding a way to properly initialize entity from data reader wasn't fun.

**Comment by fredericDelaporte** [comment]
I will not have time for reviewing until next week.

**Comment by bahusoid** [comment]
To make things more clear I decided to prepare another pull request.

**Pull request closed without merging**